spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala

```diff
@@ -69,10 +84,82 @@ private[bigdl] class BlasWrapper(val module: AbstractModule[Activity, Activity,
     }).toArray
   }
 
+  /**
+   * Blas layers may not have good performance with mkldnn backend,
+   * so we can use java multi thread to run blas layers and thread number is
+   * related with batch size and core number.
+   * input and output for this module must be in batch.
+   */
+  private def setMultiThreadEnv(input: Activity): Unit = {
+    initEnv = true
+    val multiThread = System.getProperty("multiThread", "false").toBoolean
+    if (this.train && multiThread) {
+      throw new IllegalArgumentException("Please not set multiThread to true for model training")
+    }
+    if (this.train
+      || !multiThread
+      || (_outputFormats != null && _outputFormats.length != 1)
+      || (_outputFormats != null && _inputFormats != null
+      && _inputFormats(0).shape(0) != _outputFormats(0).shape(0))
+      || !flattenInput(input)
+    ) {
+      return
+    }
+    batchSize = tensorBuffer(0).size(1)
+    val t = batchSize % Engine.coreNumber()
+    if (t != 0 || batchSize < 2 || Engine.coreNumber() < 2) {
+      logger.warn("If you want to use multiThread property to speed up, " +
+        "please attention core number should be greater than 1, " +
+        s"batch size should be greater than 1 and divided by core number, " +
+        s"but now get core number ${Engine.coreNumber()} batch size ${batchSize}")
+      return
+    }
+    subModelNumber = Engine.coreNumber()
+    initModules()
+    withMultiThread = true
+  }
+  private def flattenInput(input: Activity): Boolean = {
+    val inputDepth = if (input.isTensor) 1 else input.toTable.length()
+    if (tensorBuffer == null) tensorBuffer = new Array[Tensor[Float]](inputDepth)
+    var batch : Int = 0
+    if (inputDepth == 1) {
+      tensorBuffer(0) = input.toTensor[Float]
+    } else {
+      val in = input.toTable
+      for (i <- 1 to in.length()) {
+        if (in.get(i).get.isInstanceOf[Table]) return false
+        tensorBuffer(i - 1) = in.get[Tensor[Float]](i).get
+        if (i == 1) batch = tensorBuffer(i - 1).size(1)
+        if (batch != tensorBuffer(i - 1).size(1)
+          && !module.isInstanceOf[DetectionOutputSSD[Float]]) {
```

Comment:
Add a utility as a private class maintain a hashset for all these kinds of layers ?

Reply:
Now there is only one special layer, I think add some comments is better. If we have more such cases in the future, we can consider how to design a set to record.